sdks/java/harness/src/main/java/org/apache/beam/fn/harness/SplittableSplitAndSizeRestrictionsDoFnRunner.java

```diff
@@ -0,0 +1,481 @@
+        InputT, RestrictionT, PositionT, WatermarkEstimatorStateT, OutputT>
+    implements FnApiStateAccessor.MutatingStateContext<Object, BoundedWindow> {
+
+  /** A registrar which provides a factory to handle Java {@link DoFn}s. */
+  @AutoService(PTransformRunnerFactory.Registrar.class)
+  public static class Registrar implements PTransformRunnerFactory.Registrar {
+    @Override
+    public Map<String, PTransformRunnerFactory> getPTransformRunnerFactories() {
+      return ImmutableMap.<String, PTransformRunnerFactory>builder()
+          .put(PTransformTranslation.SPLITTABLE_SPLIT_AND_SIZE_RESTRICTIONS_URN, new Factory())
+          .build();
+    }
+  }
+
+  static class Factory implements PTransformRunnerFactory {
+
+    @Override
+    public final void addRunnerForPTransform(Context context) throws IOException {
+      addRunnerForSplitAndSizeRestriction(context);
+    }
+
+    private <InputT, RestrictionT, PositionT, WatermarkEstimatorStateT, OutputT>
+        void addRunnerForSplitAndSizeRestriction(Context context) throws IOException {
+      SplittableSplitAndSizeRestrictionsDoFnRunner<
+              InputT, RestrictionT, PositionT, WatermarkEstimatorStateT, OutputT>
+          runner =
+              new SplittableSplitAndSizeRestrictionsDoFnRunner<>(
+                  context.getPipelineOptions(),
+                  context.getPTransform(),
+                  context.getComponents(),
+                  context::getPCollectionConsumer,
+                  context.getBundleFinalizer(),
+                  FnApiStateAccessor.Factory.factoryForPTransformContext(context));
+
+      context.addPCollectionConsumer(
+          context
+              .getPTransform()
+              .getInputsOrThrow(ParDoTranslation.getMainInputName(context.getPTransform())),
+          runner::processElement);
+      context.addTearDownFunction(runner::tearDown);
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////////
+
+  private final boolean observesWindow;
+  private final PipelineOptions pipelineOptions;
+
+  private final DoFnInvoker<InputT, OutputT> doFnInvoker;
+
+  private final FnDataReceiver<
+          WindowedValue<KV<KV<InputT, KV<RestrictionT, WatermarkEstimatorStateT>>, Double>>>
+      mainOutputConsumer;
+
+  private final FnApiStateAccessor<?> stateAccessor;
+  private final SplitRestrictionArgumentProvider mutableArgumentProvider;
+
+  private final DoFnSchemaInformation doFnSchemaInformation;
+  private final Map<String, PCollectionView<?>> sideInputMapping;
+
+  private RestrictionT currentRestriction;
+  private WatermarkEstimatorStateT currentWatermarkEstimatorState;
+
+  private RestrictionTracker<RestrictionT, PositionT> currentTracker;
+
+  /**
+   * If non-null, set to true after currentTracker has had a tryClaim issued on it. Used to ignore
+   * checkpoint split requests if no progress was made.
+   */
+  private @Nullable AtomicBoolean currentTrackerClaimed;
```

Comment:
This isn't used anywhere, FnApiDoFnRunner uses the boolean of the same name to prevent splits if no claim has ever been attempted.

IIUC this isn't responding to split requests it seems like this should be removed.

Reply:
Actually it is assigned and used here: https://github.com/apache/beam/pull/34919/files#diff-c9b0240c608e2774883a6af75036bf6897e5a993afa1766a0713795a7e89f70bR239

I don't really understand _why_ we allow `@SplitRestriction` to have a current tracker parameter, but it seems we do? It doesn't look like it really responds to splits well. Certainly not the same way the `@TruncateRestriction` does...

Reply:
It is set but never read, I think it can just be removed from this class as it isn't being used to prevent splitting as in the FnApiDoFnRunner.  I added this to prevent splitting if no progress has been made, if this class isn't splitting then we don't need it.  I think I likely just added it to this method since it had a tracker and since it was difficult to tell from the previous mega-file what dofns did support splitting.

Reply:
![image](https://github.com/user-attachments/assets/626a27cc-d479-4360-aa22-52988d0b4fba)

Reply:
Do you mean that the `ClaimObserver` can be removed? Can `currentTracker` be removed and we could eliminate `RestrictionTracker` parameters from `@SplitRestriction` and `@GetSize`?

Reply:
I'm just saying to remove currentTrackerClaimed from this class. I see that we do read that it was set and that we set it to true, but nothing actually utilizes that value by reading currentTrackerClaimed.get() (where in the previous class that was used to prevent splitting).  So we can just remove it.

Reply:
Ah I see what you mean. Done. Based on that, does it even make sense to set `currentTracker = new tracker`. We ignore it afterwards, and never react to touching it.

Reply:
Not only in this class. But this line https://github.com/apache/beam/blob/298e0e0b9096125dfbfd58231b445deec502122b/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java#L231 would appear to allow the parameter, but then we just throw it away and AFAICT there's no way any other thread could interact with it.